pkg/edgehub/clients/wsclient/websocket_test.go

```diff
@@ -0,0 +1,352 @@
+func init() {
+	newTestServer()
+
+	err := util.GenerateTestCertificate("/tmp/", "edge", "edge")
+	if err != nil {
+		panic("Error in creating fake certificates")
+	}
+}
+
+//Message is content object to be passed in as message object
+type Message struct {
+	Name string `json:"name"`
+}
+
+var testServer *httptest.Server
+var upgrader websocket.Upgrader
+
+//newTestServer() starts a fake server for testing
+func newTestServer() {
+	testServer = httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		switch {
+		case strings.Contains(r.RequestURI, "/normal"):
+			c, err := upgrader.Upgrade(w, r, nil)
+			if err != nil {
+				return
+			}
+			defer c.Close()
+			m := model.Message{}
+			for {
+				err := c.ReadJSON(&m)
+				if err != nil {
+					break
+				}
+				err = c.WriteJSON(m)
+				if err != nil {
+					break
+				}
+			}
+
+		case strings.Contains(r.RequestURI, "/bad_request"):
+			w.WriteHeader(http.StatusBadRequest)
+
+		case strings.Contains(r.RequestURI, "/wrong_send"):
+			c, err := upgrader.Upgrade(w, r, nil)
+			if err != nil {
+				return
+			}
+			defer c.Close()
+			m := model.Message{}
+			for {
+				err := c.ReadJSON(&m)
+				if err != nil {
+					break
+				}
+				err = c.WriteMessage(3, []byte(""))
+				if err != nil {
+					break
+				}
+			}
+		}
+	}))
+}
+
+//TestNewWebSocketClient tests the NewWebSocketClient function that creates the WebSocketClient object
+func TestNewWebSocketClient(t *testing.T) {
+	tests := []struct {
+		name string
+		conf *WebSocketConfig
+		want *WebSocketClient
+	}{
+		{"TestNewWebSocketClient: ", &WebSocketConfig{
+			URL:              "ws://" + testServer.Listener.Addr().String() + "/normal",
+			CertFilePath:     "/tmp/edge.crt",
+			KeyFilePath:      "/tmp/edge.key",
+			HandshakeTimeout: 500 * time.Second,
+			WriteDeadline:    100 * time.Second,
+			ReadDeadline:     100 * time.Second,
+			ExtendHeader:     http.Header{},
+		}, &WebSocketClient{
+			config: &WebSocketConfig{
+				URL:              "ws://" + testServer.Listener.Addr().String() + "/normal",
+				CertFilePath:     "/tmp/edge.crt",
+				KeyFilePath:      "/tmp/edge.key",
+				HandshakeTimeout: 500 * time.Second,
+				WriteDeadline:    100 * time.Second,
+				ReadDeadline:     100 * time.Second,
+				ExtendHeader:     http.Header{},
+			},
+		}},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if got := NewWebSocketClient(tt.conf); !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("NewWebSocketClient() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+//TestInit tests the procurement of the WebSocketClient
+func TestInit(t *testing.T) {
+	tests := []struct {
+		name          string
+		fields        WebSocketClient
+		expectedError error
+	}{
+		{name: "TestInit: Success in connection ",
+			fields: WebSocketClient{
+				webConn:  &websocket.Conn{},
+				sendLock: sync.Mutex{},
+				config: &WebSocketConfig{
```

Comment:
the creation of the websocketconfig object can be a separate method with parameters  , as all tests create this.